ntex-bytes/src/buf/pool.rs

```diff
@@ -0,0 +1,269 @@
+use bytes::{Buf, BufMut, Bytes, BytesMut, buf::UninitSlice};
+use std::future::{self, Future};
+use std::io::IoSlice;
+use std::ops::Deref;
+use std::rc::Rc;
+use std::sync::Arc;
+
+/// A source of buffers.
+///
+/// Dropped buffers will automatically return to this pool.
+pub trait BufferSource {
+    /// The type of buffer returned by this buffer source that owns its contents.
+    type Owned: Owned;
+
+    /// The type of accumulator of buffers used by encoders.
+    type Accumulator: BufferAccumulator<Shared = <Self::Owned as Owned>::Shared>;
+
+    /// The type of future returned by the [`BufferSource::take`] impl.
+    type TakeFuture: Future<Output = Self::Owned>;
```

Comment:
should we support allocation failures?

Reply:
If the buffer pool cannot return a buffer because there isn't one, then the future doesn't resolve until it can.

Reply:
so potentially list of waiters is unlimited?

Reply:
Yes. In our implementation it's a `slab::Slab<Option<std::task::Waker>`. Each entry in the slab is <= 16 bytes. I don't think we're worried that the slab itself will become large enough that we need to worry about its memory usage.